### PR TITLE
RSC-1224-Load Error - convert tests to RTL

### PR DESCRIPTION
**Pull request by Serhii-Bondarchuk**
Ticket link: https://issues.sonatype.org/browse/RSC-1224
Jenkins link:

### CODE REVIEW

File: lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx [hunk 1]
```diff
@@ -4,86 +4,99 @@
...
-        .toContain('An error occurred loading data');
+    const el = renderEl({ error: '' });
+    expect(el?.textContent).toContain('An error occurred loading data. ');
   });
 
   it('uses the specified title message instead of the default', function() {
-    expect(getShallowComponent({ error: 'Error!', titleMessage: 'This is bad' }).render().text())
-        .not.toContain('An error occurred loading data');
+    const el = renderEl({ error: '', titleMessage: 'This is bad!' });
+    expect(el?.textContent).not.toBe('An error occurred loading data. ');
```
Comment (rpokorny) [inline lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx:31]:
```suggestion
    expect(el?.textContent).not.toContain('An error occurred loading data. ');
```

File: lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx [hunk 2]
```diff
@@ -4,86 +4,99 @@
...
+    expect(el?.textContent).toContain('This is bad! ');
 
-    expect(getShallowComponent({ error: 'Error!', titleMessage: 'This is bad' }).render().text())
-        .toContain('This is bad');
   });
 
   it('Uses the error as children', function() {
```
Comment (rpokorny) [inline lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx:36]: "Children" isn't really a DOM concept, it's a react concept.
```suggestion
  it('renders the error as part of the text content', function() {
```

File: lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx [hunk 3]
```diff
@@ -4,86 +4,99 @@
...
   it('renders a retry button if there is an error and retryHandler is set', function() {
-    expect(getShallowComponent({ error: 'Error!' })).not.toContainMatchingElement(NxButton);
-
-    expect(getShallowComponent({ error: 'Error!', retryHandler: () => {} })).toContainMatchingElement(NxButton);
-    expect(getShallowComponent({ error: 'Error!', retryHandler: () => {} }).find(NxButton)).toIncludeText('Retry');
+    const elWithoutRetryButton = renderEl({ error: 'Error' });
+    expect(elWithoutRetryButton?.textContent).not.toContain('Retry');
+    const elRetryButton = renderEl({ error: 'Error', retryHandler: () => {} });
+    const retryButton = screen.getByRole('button', {name: 'Retry'});
+    expect(elRetryButton?.textContent).toContain('An error occurred loading data. Error');
```
Comment (rpokorny) [inline lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx:51]: This line doesn't need to be part of this test

File: lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx [hunk 4]
```diff
@@ -4,86 +4,99 @@
...
+    expect(retryButton).toBeInTheDocument();
```
Comment (rpokorny) [inline lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx:52]:
```suggestion
    expect(quickRender({ error: 'Error', retryHandler: () => {} }).getByRole('button', { name: 'Retry' })).toBeInTheDocument();
    const elRetryButton = renderEl({ error: 'Error', retryHandler: () => {} });
    const retryButton = screen.getByRole('button', {name: 'Retry'});
    expect(elRetryButton?.textContent).toContain('An error occurred loading data. Error');
    expect(retryButton).toBeInTheDocument();
```

`quickRender` being a function that we usually define alongside `renderEl`, using our `rtlRender` utility function

File: lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx [hunk 5]
```diff
@@ -4,86 +4,99 @@
...
   it('adds the appropriate class, variant, and type to the retry button', function() {
-    const button = getShallowComponent({ error: 'Error!', retryHandler: () => {} }).find(NxButton);
-
-    expect(button).toHaveClassName('nx-load-error__retry');
-    expect(button).toHaveProp('variant', 'error');
-    expect(button).toHaveProp('type', 'button');
+    renderEl({ error: 'Error', retryHandler: () => {} });
+    const retryButton = screen.getByRole('button', {name: 'Retry'});
+    expect(retryButton).toBeInTheDocument();
+    expect(retryButton).toHaveClass('nx-load-error__retry');
```
Comment (rpokorny) [inline lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx:59]: We consider class names to be implementation details.  The relevant user-facing effect, the styling, is tested via visual tests

File: lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx [hunk 6]
```diff
@@ -4,86 +4,99 @@
...
+    const retryButton = screen.getByRole('button', {name: 'Retry'});
+    expect(retryButton).toBeInTheDocument();
+    await user.click(retryButton);
     expect(retryHandler).toHaveBeenCalled();
   });
 
-  it('passes unknown props to the NxErrorAlert element', function() {
-    const onClick = jest.fn(),
-        component = getShallowComponent({ error: 'err', id: 'foo', onClick });
+  it('calls the onClose function when the "X" button is clicked (to dismiss alert)', async function() {
```
Comment (rpokorny) [inline lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx:81]: I'd break this down a bit.  I'd create a `describe` block for "when onClose is defined" and they I'd have two tests: one that tests that that causes the Close button to be rendered (and that the Close button is not rendered when `onClose` is not defined), and a second test that tests that clicking the Close button fires `onClose`

File: lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx [hunk 7]
```diff
@@ -4,86 +4,99 @@
...
+    expect(onClose).not.toHaveBeenCalled();
+    const closeButton = screen.getByRole('button', { name: 'Close' });
+    expect(closeButton).toBeInTheDocument();
+    await user.click(closeButton);
+    expect(onClose).toHaveBeenCalled();
+  });
 
-    expect(component).toHaveProp('id', 'foo');
-    expect(component).toHaveProp('onClick', onClick);
+  it('passes unknown props to the NxErrorAlert element', async function() {
```
Comment (rpokorny) [inline lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx:92]:
```suggestion
  it('passes unknown props to the alert element', async function() {
```

File: lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx (hunk 6 above)
Comment (Serhii-Bondarchuk) [reply lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx]: > when onClose is defined

Done 9c3a7b928eb032da2bbd061cecf1ba01a191e71b

File: lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx (hunk 3 above)
Comment (Serhii-Bondarchuk) [reply lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx]: > This line doesn't need to be part of this test

Can you, please, specify what kind of line you mean?  Is it something wrong in this code?

Comment (Serhii-Bondarchuk) [reply lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx]: > This line doesn't need to be part of this test

And I change it previous comment, is it fixed now? ```
it('renders a retry button if there is an error and retryHandler is set', async function() {
    const elWithoutRetryButton = renderEl({ error: 'Error' });
    expect(elWithoutRetryButton?.textContent).not.toContain('Retry');
    expect(quickRender({ error: 'Error', retryHandler: () => {} }).getByRole('button', { name: 'Retry' }))
        .toBeInTheDocument();
  });
```

Comment (Serhii-Bondarchuk) [reply lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx]: > This line doesn't need to be part of this test

And I change it previous comment, is it fixed now? 
```
it('renders a retry button if there is an error and retryHandler is set', async function() {
const elWithoutRetryButton = renderEl({ error: 'Error' });
expect(elWithoutRetryButton?.textContent).not.toContain('Retry');
expect(quickRender({ error: 'Error', retryHandler: () => {} }).getByRole('button', { name: 'Retry' }))
.toBeInTheDocument();
});
```

File: lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx [hunk 8]
```diff
@@ -4,86 +4,99 @@
...
+    const user = userEvent.setup();
+    const onClick = jest.fn();
+    renderEl({ error: 'err', id: 'foo', onClick });
+    expect(onClick).not.toHaveBeenCalled();
+    const alert = screen.getByRole('alert');
+    expect(alert).toHaveAttribute('id', 'foo');
+    await user.click(alert);
+    expect(onClick).toHaveBeenCalled();
   });
```
[the remark this answers is not in the document]
Comment (Serhii-Bondarchuk) [reply lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx:112]: > Also check that extra classnames specified in props are handled correctly. See https://github.com/sonatype/sonatype-react-shared-components/blob/main/lib/src/components/NxToast/__tests__/NxToast.test.tsx#L37 for how we usually do that

Done a40eb321cc31797ebc027a9d71ec9f0718e5028c

File: lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx (hunk 5 above)
Comment (Serhii-Bondarchuk) [reply lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx]: > We consider class names to be implementation details. The relevant user-facing effect, the styling, is tested via visual tests

Done 3e18fa6c54f37ee9e88ac09c178121d805dcaa89

File: lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx [hunk 9]
```diff
@@ -4,86 +4,109 @@
...
+    const el = renderEl({ error: 'Server Error', titleMessage: 'This is bad!' });
+    expect(el?.textContent).toEqual('This is bad! Server Error');
   });
 
-  it('renders a retry button if there is an error and retryHandler is set', function() {
-    expect(getShallowComponent({ error: 'Error!' })).not.toContainMatchingElement(NxButton);
-
-    expect(getShallowComponent({ error: 'Error!', retryHandler: () => {} })).toContainMatchingElement(NxButton);
-    expect(getShallowComponent({ error: 'Error!', retryHandler: () => {} }).find(NxButton)).toIncludeText('Retry');
+  it('renders a retry button if there is an error and retryHandler is set', async function() {
```
Comment (avnvbhatta) [inline lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx:47]: Doesn't look like this is an `async` function?

File: lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx [hunk 10]
```diff
@@ -4,86 +4,109 @@
...
+    const elWithoutRetryButton = renderEl({ error: 'Error' });
+    expect(elWithoutRetryButton?.textContent).not.toContain('Retry');
+    expect(quickRender({ error: 'Error', retryHandler: () => {} }).getByRole('button', { name: 'Retry' }))
+        .toBeInTheDocument();
   });
 
   it('adds the appropriate class, variant, and type to the retry button', function() {
```
Comment (avnvbhatta) [inline lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx:54]: This test isn't testing anything for the class or variant. The wording of the `it` block needs to be changed.

File: lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx [hunk 11]
```diff
@@ -4,86 +4,109 @@
...
-    const button = getShallowComponent({ error: 'Error!', retryHandler: () => {} }).find(NxButton);
-
-    expect(button).toHaveClassName('nx-load-error__retry');
-    expect(button).toHaveProp('variant', 'error');
-    expect(button).toHaveProp('type', 'button');
+    renderEl({ error: 'Error', retryHandler: () => {} });
+    const retryButton = screen.getByRole('button', {name: 'Retry'});
```
Comment (avnvbhatta) [inline lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx:56]:
```suggestion
    const retryButton = screen.getByRole('button', { name: 'Retry' });
```

File: lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx [hunk 12]
```diff
@@ -4,86 +4,109 @@
...
+    expect(retryButton).toBeInTheDocument();
+    expect(retryButton).toHaveAttribute('type', 'button');
   });
 
   it('sets the retry button type to "submit" when submitOnRetry is set to true', function() {
-    const props = { error: 'Error!', submitOnRetry: true };
-    const button = getShallowComponent(props).find(NxButton);
-    expect(button).toHaveProp('type', 'submit');
+    renderEl({ error: 'Error', submitOnRetry: true });
+    const retryButton = screen.getByRole('button', {name: 'Retry'});
```
Comment (avnvbhatta) [inline lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx:63]:
```suggestion
    const retryButton = screen.getByRole('button', { name: 'Retry' });
```

File: lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx [hunk 13]
```diff
@@ -4,86 +4,109 @@
...
-
+  it('calls the retryHandler when the retry button is clicked', async function() {
+    const user = userEvent.setup();
+    const retryHandler = jest.fn();
+    renderEl({ error: 'Error!', retryHandler });
     expect(retryHandler).not.toHaveBeenCalled();
-
-    component.find(NxButton).simulate('click');
-
+    const retryButton = screen.getByRole('button', {name: 'Retry'});
```
Comment (avnvbhatta) [inline lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx:73]:
```suggestion
    const retryButton = screen.getByRole('button', { name: 'Retry' });
```

File: lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx [hunk 14]
```diff
@@ -4,86 +4,109 @@
...
+    expect(retryButton).toBeInTheDocument();
+    await user.click(retryButton);
     expect(retryHandler).toHaveBeenCalled();
   });
 
-  it('passes unknown props to the NxErrorAlert element', function() {
-    const onClick = jest.fn(),
-        component = getShallowComponent({ error: 'err', id: 'foo', onClick });
+  describe('when onClose is defined', ()=> {
+    it('if onClose function is defined then the "X" button is appear', function() {
```
Comment (avnvbhatta) [inline lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx:80]:
```suggestion
    it('the "X" button appears', function() {
```
The `describe` block already has "when onClose is defined"

File: lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx (hunk 3 above)
Comment (rpokorny) [reply lib/src/components/NxLoadError/__tests__/NxLoadError.test.tsx]: > And I change it previous comment, is it fixed now?

That seems right.  My point was that checking the text content of `elRetryButton` seemed unnecessary

**Review by rpokorny — APPROVED** [review]

**Review by avnvbhatta — APPROVED** [review]

**Pull request merged**